Spartan Laser + Spartan Shield sounds + Misc Armour changes

Stingray540 opened the pull request:
Essentially, adding this PR in order to get help/comments.

-Correctly links the Spartan Laser to the sprite already present in the files
-Adds a variable for if a weapon needs to be "charged", defaulted to 0 for everything aside from/charged weapons
-Adds a delay to firing the Spartan Laser, as previously at was instant fire. The sound for firing is linked to the charge, however, I'm currently trying to cause the do_after to be triggered at the same time as the charge/fire sound.

Most of this is likely incomplete and inefficient as I'm very new to this, so please give any tips you can think of. Thanks!

Stingray540 committed:
Start of properly implementing spartan lasers.

Stingray540 committed:
More splaser changes, still not finished.

Stingray540 committed:
-Causes the charge to actually break when the do_after fails
-Correctly syncs the length of the charge to the length of the sound .ogg
-Causes the .ogg to play when the charge starts rather than when the weapon actually fires

caelaislinn commented on code/modules/halo/weapons/projectiles/beams.dm:
```diff
@@ -1,7 +1,7 @@
 	name = "spartan laser"
 	icon_state = "heavylaser"
-	damage = 100
+	damage = 200
```
This could potentially be boosted. We don't have a unified way of handling vehicles but 200 is around the max HP of mobs which means there's a small chance of survival if being hit

caelaislinn commented on code/modules/projectiles/gun.dm:
```diff
@@ -228,6 +231,7 @@
 		O.emp_act(severity)
 
 /obj/item/weapon/gun/afterattack(atom/A, mob/living/user, adjacent, params)
+
```
Unnecesary line change

caelaislinn commented on code/modules/projectiles/gun.dm:
```diff
@@ -239,6 +243,13 @@
 		to_chat(user, "<span class='warning'>You refrain from firing your [src] as your intent is set to help.</span>")
+
+	if(is_charged_weapon==1)
+		playsound(src.loc, charge_sound, 100, 1)
```
This will play the firing sound I believe?

Stingray540 replied:
Yes. Currently, it'll play the charging/firing sound (which is one .ogg). Is it possible for me to cause the .ogg to stop playing if the do_after fails, or will I have to split the actual laser firing sound and the laser charging sound into two in order to prevent it triggering even when not actually fired?

Stingray540 replied:
As of the latest commit, I'd say I've got three main things to sort:

-Cause the charging sound to stop when charging fails rather than still play the entire .ogg

-Add some kind of tracer during the charging phase to alert enemies due to its strength. No idea how to do this. Not required, but unless it stays admin-spawn forever will likely be needed for balance.

-Get a l/r hand sprite for the weapon. I'm not a spriter so I can't do this, but I guess I'll try and lean if I complete everything else.

Thanks for all the help so far.

caelaislinn replied:
How long is the ogg? It is probably fine to just let it play out then only play the fire sound if the do_after is successful. You could probably modify or reuse the halo rocket launcher in hand sprite. 

A tracer for the laser while charging will need some finicking but its doable

Stingray540 replied:
Currently the charge .ogg is around 3 seconds and the fire .ogg around 1, making the total around 4 when firing is successful. Your suggestion is essentially what I've done - the fire sound is added to the beam itself firing and the charge effect will play just before the do_after starts. (Being 3s long makes it match). And yeah, the tracers are going to be a pain for me. Some help with that would be great. Thanks again, Cael.

caelaislinn commented on code/modules/projectiles/gun.dm:
```diff
@@ -87,6 +87,9 @@
 	var/list/attachments_on_spawn = list()
 
+	var/is_charged_weapon = 0 //Does the weapon require charging? Defaults to 0 unless it's from /charged weapon sets
+	var/arm_time = 2.5 //Default charge time for weapons that charge
```
Time in byond is measured in deciseconds, so this will amount to 1/4 of a second. To make it 2.5 seconds you should either set this value to 25 or you could go var/arm_time = 2.5 SECONDS

SECONDS is a macro which automatically converts a value from deciseconds to seconds by multiplying it by 10

caelaislinn reviewed (changes requested):
Ignore the travis failure. This may be good to merge, but I'd like to manually test it first

Stingray540 committed:
-Removes the default beam sound from triggering as well as the spartan laser sound

-Tweaks the default arm_time of charged weapons to 25 from 2.5 since I didn't realise it was in deciseconds

Delnth commented:
There is an item in the code that's called >Infrared emitter.

This item offers an option to render the infrared beam visible or not. You might want to check it out and work from there. It has a distance limit, though.

Stingray540 committed:
-Adds two sound files (Spartan_Laser_Beam_Shot_Sound_Effect.ogg and Spartan_Laser_Charge_Sound_Effect.ogg)

-These sound fires are effectively just the original firing sound split on two, one assigned to the charge, the other to the actual beam firing sound.

Stingray540 commented:
As of the latest commit, I have two more main goals, one of which I'm not entirely sure I can complete on my own after multiple attempts (at my beginner level of coding for SS13).

-Add the IR beam effect to the spartan laser while it is charging. I've messed around with this a bunch and can't quite figure it out. This isn't strictly an essential feature, but for balance it is. (The instadeath beam needs *some* tells, after all). If any actual developers would be willing to either 'handhold' me through it so I learn, or take over for this feature, I'd be very grateful.

-Make the 'charging' part of the sound stop once the charging do_after is stopped. The most recent commit prevents the 'beam' sound from playing by splitting the file in two, but if possible I'd still like it to stop the .ogg entirely, as currently, it will still play the entire charging sound even if, say, it is cancelled as soon as it begins. Not even sure if .ogg can be stopped midway.

The weapon also needs l/r hand sprites. I'm not a spriter so I'm not considering this one of my personal goals, it could be merged without it, however.

Thank you for all the help so far.

Stingray540 committed:
-Adds thickness to spartan armour (now equal to colossus)

-Makes URFC armour spaceproof

-Adds a warning siren and shield recharge sound which triggers when shields are at 0 and shields begin to recharge respectively

Stingray540 committed:
Splaser sprite changes

Stingray540 committed:
Merge pull request #1 from Stingray540/Armour_Changes

Spartan Laser / Armour Changes

Stingray540 commented:
Edited this PR to include my armour changes - which adds sounds to spartan shield breaking and charging, along with thickness changes, bug fixes when spawning. Also spaceproofs the URFC armour
